File: Java/src/com/mergesort/MtMergeSort.java
package com.mergesort;

import java.util.concurrent.RecursiveAction;

public class MtMergeSort extends RecursiveAction {
  private Integer[] arr;
  private int start;
  private int end;

  MtMergeSort(Integer[] arr, int start, int end) {
    this.arr = arr;
    this.start = start;
    this.end = end;
  }

  @Override
  protected void compute() {
    if (start < end) {
      int mid = start + (end - start) / 2;
      invokeAll(new MtMergeSort(arr, start, mid), new MtMergeSort(arr, mid + 1, end));
      MergeUtils.merge(arr, start, mid, end);
    }
  }
}

File: C++/src/mt_merge_sort.cpp
#include "mt_merge_sort.hpp"
#include "merge_utils.hpp"
#include <thread>

void MtMergeSort::merge_sort(std::vector<int> &vec, size_t start, size_t end)
{
  if (start < end)
  {
    size_t mid = (start + end) / 2;

    std::thread left(MtMergeSort::merge_sort, std::ref(vec), start, mid);
    std::thread right(MtMergeSort::merge_sort, std::ref(vec), mid + 1, end);

    left.join();
    right.join();

    MergeUtils::merge(vec, start, mid, end);
  }
}